### PR TITLE
Refactor: move into package + add caching of files

### PR DESCRIPTION
**Pull request by RDIL**
I used the namespace `me.param` since that is your domain
local module names should be uppercase
### also this may break stuff

### CODE REVIEW

**Comment by RDIL** [comment]
@paramt may want to remove the requirement for test job to pass

**Comment by RDIL** [comment]
Okay done

**Comment by paramt** [comment]
I apologize for my ignorance, but I'm having a hard time understanding the file caching logic and how it helps... [Also, as far as I'm aware, all python imports are lower-case](https://www.python.org/dev/peps/pep-0008/#package-and-module-names).

**Comment by RDIL** [comment]
Cache allows for the code to resolve the file it is reading/writing from faster, to speed up the process

File: src/app.py [hunk 1]
```diff
@@ -4,14 +4,21 @@
 import time
 import algorithm
 import constants
+import cache
 
-reddit = praw.Reddit('MemeAdviser')
+# check if this is being imported or not
+if __name__ not "__main__":
```
Comment (paramt) [inline src/app.py]:
```suggestion
if __name__ != "__main__":
```
`SyntaxError: invalid syntax`

File: src/app.py [hunk 2]
```diff
@@ -4,14 +4,21 @@
...
+    raise OSError("You need to run this file directly, not import it silly!")
 
-with open("../replied.txt", "r") as f:
-    replied = list(filter(None, f.read().split("\n")))
+reddit = praw.Reddit('MemeAdviser')
 
-with open("../subscribed.txt", "r") as f:
-    subscribed = list(filter(None, f.read().split("\n")))
+replied_file = Cache.AbstractFile("../replied.txt")
```
Comment (paramt) [inline src/app.py]:
```suggestion
replied_file = cache.AbstractFile("../replied.txt")
```

File: src/app.py [hunk 3]
```diff
@@ -4,14 +4,21 @@
...
+subscribed_file = Cache.AbstractFile("../subscribed.txt")
```
Comment (paramt) [inline src/app.py]:
```suggestion
subscribed_file = cache.AbstractFile("../subscribed.txt")
```

File: src/app.py [hunk 4]
```diff
@@ -4,14 +4,21 @@
...
+fileLoadCache = Cache.FileCache()
```
Comment (paramt) [inline src/app.py]:
```suggestion
fileLoadCache = cache.FileCache()
```

File: src/cache.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+# |==========================================|
+# | Copyright (c) 2017-present Reece Dunham  |
+# | All rights reserved                      |
+# | Used with explicit permission            |
+# |==========================================|
+
+# Taken from a project that may become open source soon, please keep the copyright header
+
```
Comment (paramt) [inline src/cache.py:10]:
```suggestion

from io import TextIOWrapper

```

`NameError: name 'TextIOWrapper' is not defined`

File: src/cache.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    def add(self, i: object) -> bool:
+        try:
+            self.file_caches.append(i)
+        except:
+            return False
+        finally:
+            return True
+
+    def clear_file(file: str) -> None:
```
Comment (paramt) [inline src/cache.py]:
```suggestion
    def clear_file(self, file: str) -> None:
```
`TypeError: clear_file() takes 1 positional argument but 2 were given`

File: src/app.py [hunk 7]
```diff
@@ -34,9 +41,8 @@
     try:
         # Update replied.txt
         replied.append(submission.id)
-        with open("../replied.txt", "w") as f:
-            for post_id in replied:
-                f.write(post_id + "\n")
+        for post_id in replied:
+            replied.write_to_file(post_id + "\n")
```
Comment (paramt) [inline src/app.py]:
```
Traceback (most recent call last):
  File "app.py", line 45, in <module>
    replied.write_to_file(post_id + "\n")
AttributeError: 'list' object has no attribute 'write_to_file'
```

**Review by paramt — CHANGES_REQUESTED** [review]
The script isn't updating `replied.txt`, which needs to be fixed before merging. I've highlighted several minor syntax errors as well, but other than that it looks good. After you apply the changes, I'll have to test it again before merging the PR

File: src/cache.py (hunk 5 above)
Comment (RDIL) [reply src/cache.py:10]: Huh, should be in memory by default

File: src/app.py (hunk 7 above)
Comment (RDIL) [reply src/app.py]: Ah, I can fix that.

File: src/cache.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        self.file_caches = []
+
+    def refresh(self, file: AbstractFile) -> (object, list):
+        """
+        Refreshes the @TextIOWrapper for the file
+
+        :param file: the @AbstractFile to refresh
+        :return: some list thing
+        """
+        with open(file.get_name(), mode="r") as filehandler:
```
Comment (paramt) [inline src/cache.py:51]:
```
Traceback (most recent call last):
  File "app.py", line 18, in <module>
    replied = fileLoadCache.refresh(replied_file.get_name())
  File "cache.py", line 51, in refresh
    with open(file.get_name(), mode="r") as filehandler:
AttributeError: 'str' object has no attribute 'get_name'
```

**Review by paramt — CHANGES_REQUESTED** [review]

File: src/cache.py (hunk 8 above)
Comment (RDIL) [reply src/cache.py:51]: oh right smh

Comment (RDIL) [reply src/cache.py:51]: which python version?

Comment (paramt) [reply src/cache.py:51]: 2.7, but I can upgrade if it's required

Comment (RDIL) [reply src/cache.py:51]: sir
your going to need latest

Comment (paramt) [reply src/cache.py:51]: I can run 3.6.7

Comment (RDIL) [reply src/cache.py:51]: 3.7.2 is prob best

Comment (paramt) [reply src/cache.py]: Hmm.. would 3.7.1 work?

Comment (paramt) [reply src/cache.py:51]: @RDIL 3.7.2 is now installed on the server

Comment (RDIL) [reply src/cache.py:51]: Yeah that should work - thanks!

**Comment by RDIL** [comment]
Anything else not working?

**Comment by paramt** [comment]
@RDIL Unfortunately, I won't be able to test the code until Sunday evening

**Comment by RDIL** [comment]
No worries

File: src/app.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        minutes = str(minutes // 60) + "h " + str(minutes % 60) + "min"
+    else:
+        minutes = str(round((time.time() - submission.created_utc) / 60)) + " minutes"
+
+    if submission.id not in replied:
+        try:
+            # Update replied.txt
+            replied.append(submission.id)
+            for post_id in replied:
+                replied.write_to_file(post_id + "\n")
```
Comment (paramt) [inline src/app.py:42]:
```
Traceback (most recent call last):
  File "app.py", line 97, in <module>
    main(constants.Thresholds)
  File "app.py", line 42, in main
    replied.write_to_file(post_id + "\n")
AttributeError: 'list' object has no attribute 'write_to_file'
```

**Review by paramt — CHANGES_REQUESTED** [review]

File: src/app.py (hunk 9 above)
Comment (RDIL) [reply src/app.py:42]: Something is happening with a variable conflict

Comment (paramt) [reply src/app.py:42]: I can't seem to find any conflicts... `replied` is only declared once, on line 15
It seems that `write_to_file` is a method of the `FileCache` class, but 
```python
replied = fileLoadCache.refresh(replied_file)
``` 

returns a list

Comment (RDIL) [reply src/app.py:42]: Ah, that’s it.  Any idea how to work around this?

Comment (paramt) [reply src/app.py:42]: I don't really understand `cache.py` but we could replace '`replied.write_to_file()` with 

```python 
with open("/data/replied.txt", "w") as f: 
    for post_id in replied: 
        f.write(post_id + "\n")
```

like it was before... Or does that just make the whole thing pointless?

Comment (RDIL) [reply src/app.py:42]: I need to take a break from this PR for my mental health, I’ll reopen it once I have a fix.

Comment (paramt) [reply src/app.py:42]: Take your time, don't stress about it

**Pull request closed without merging**